vpnglean/glean_parser_ext/cpp.py

```diff
@@ -0,0 +1,136 @@
+        return "Labeled<impl::{}Metric>".format(class_name)
+    generate_enums = getattr(obj, "_generate_enums", [])  # Extra Keys? Reasons?
+    if len(generate_enums):
+        for name, suffix in generate_enums:
+            if not len(getattr(obj, name)) and suffix == "Keys":
+                return util.Camelize(obj.type) + "Metric<NoExtraKeys>"
+            else:
+                # we always use the `extra` suffix,
+                # because we only expose the new event API
+                suffix = "Extra"
+                return "{}Metric<{}>".format(
+                    util.Camelize(obj.type), util.Camelize(obj.name) + suffix
+                )
+    return util.Camelize(obj.type) + "Metric"
+
+
+def extra_type_name(typ: str) -> str:
+    """
+    Returns the corresponding Rust type for event's extra key types.
+    """
+
+    if typ == "boolean":
+        return "bool"
+    elif typ == "string":
+        return "QString"
+    elif typ == "quantity":
+        return "int"
+    else:
+        return "UNSUPPORTED"
+
+
+def output_cpp(objs, output_fd, options={}):
+    """
+    Given a tree of objects, output C++ code to the file-like object `output_fd`.
+
+    :param objs: A tree of objects (metrics and pings) as returned from
+    `parser.parse_objects`.
+    :param output_fd: Writeable file to write the output to.
+    :param options: options dictionary.
+    """
+
+    # Monkeypatch a util.snake_case function for the templates to use
+    util.snake_case = lambda value: value.replace(".", "_").replace("-", "_")
+    # Monkeypatch util.get_jinja2_template to find templates nearby
+
+    def get_local_template(template_name, filters=()):
+        env = jinja2.Environment(
+            loader=jinja2.PackageLoader("cpp", "templates"),
+            trim_blocks=True,
+            lstrip_blocks=True,
+        )
+        env.filters["camelize"] = util.camelize
+        env.filters["Camelize"] = util.Camelize
+        for filter_name, filter_func in filters:
+            env.filters[filter_name] = filter_func
+        return env.get_template(template_name)
+
+    util.get_jinja2_template = get_local_template
+    get_metric_id = generate_metric_ids(objs)
+    get_ping_id = generate_ping_ids(objs)
+
+    if "pings" in objs:
+        template_filename = "cpp_pings.jinja2"
+        if objs.get("tags"):
+            del objs["tags"]
+    else:
+        template_filename = "cpp.jinja2"
+        objs = get_metrics(objs)
+
+    template = util.get_jinja2_template(
+        template_filename,
+        filters=(
+            ("cpp", cpp_datatypes_filter),
+            ("snake_case", util.snake_case),
+            ("type_name", type_name),
+            ("extra_type_name", extra_type_name),
+            ("metric_id", get_metric_id),
+            ("ping_id", get_ping_id),
+            ("Camelize", util.Camelize),
+        ),
+    )
+
+    output_fd.write(template.render(all_objs=objs))
+    output_fd.write("\n")
```

Comment:
Heh, I was gonna say. I am still happy to file tickets, I'll do that on the Glean: SDK Bugzilla component. Let me know if that is not the one for FOG.

Reply:
The transformation is not as simple as `all the nsCString replaced with QString`, but I'll definitely add a comment to all of these files.

Reply:
Ah, also. All the changes I made to glean_parser_ext are in this commit: https://github.com/mozilla-mobile/mozilla-vpn-client/pull/4919/commits/7831353767c8b8e82d4362e87e9b2765958a3a70. I commited the raw copy paste so it would be easier to see exactly what changed.

Reply:
FOG's component is `Toolkit :: Telemetry`

The reason I mention the whole "add a comment" thing is because you're likely gonna have to do this again and it might be helpful to have some documentation about what's been done.